src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs

```diff
@@ -24,6 +24,12 @@ public Task<ErrorHandleResult> Invoke(ErrorContext errorContext)
         {
             var recoveryAction = recoverabilityPolicy(configuration, errorContext);
 
+            if (recoveryAction is Discard discard)
+            {
+                Logger.Info($"Discarding message with id '{errorContext.Message.MessageId}'.{(string.IsNullOrEmpty(discard.Reason) ? string.Empty : $" Reason: {discard.Reason}")}", errorContext.Exception);
```

Comment:
If we added the guard I mentioned, this could be simplified.

Reply:
Done